flink-streaming-java/src/main/java/org/apache/flink/streaming/api/collector/selector/SelectedOutputsCollectorImpl.java

```diff
@@ -0,0 +1,62 @@
+package org.apache.flink.streaming.api.collector.selector;
+
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.Map;
+
+/**
+ * The selected outputs collector will send records to the default output,
+ * and output matching outputNames.
+ *
+ * @param <T> The type of the elements that can be emitted.
+ */
+public class SelectedOutputsCollectorImpl<T> implements SelectedOutputsCollector<T> {
+
+	private final Output<StreamRecord<T>>[] selectAllOutputs;
+	private final Map<String, Output<StreamRecord<T>>[]> outputMap;
+
+	private final boolean objectReuse;
+
+	public SelectedOutputsCollectorImpl(
+		Output<StreamRecord<T>>[] selectAllOutputs,
+		Map<String, Output<StreamRecord<T>>[]> outputMap,
+		boolean objectReuse) {
+		this.selectAllOutputs = selectAllOutputs;
+		this.outputMap = outputMap;
+		this.objectReuse = objectReuse;
+	}
+
+	@Override
+	public boolean collect(Iterable<String> outputNames, StreamRecord<T> record) {
+		boolean emitted = false;
+
+		if (selectAllOutputs.length > 0) {
+			collect(selectAllOutputs, record);
+			emitted = true;
+		}
+
```

Comment:
In the old implementation via `set`, even if the same `output` appears multiple times in `outputNames`, it will only be sent once. Now it will send multiple times and I am not sure if this behavior is correct.